charts/newrelic-infrastructure-v3/templates/_helpers_compatibility.tpl

```diff
@@ -0,0 +1,79 @@
+{{/*
+Returns true if .Values.ksm.enabled is true and the legacy disableKubeStateMetrics is not set
+*/}}
+{{- define "newrelic.compatibility.ksm.enabled" -}}
+{{- if and .Values.ksm.enabled (not .Values.disableKubeStateMetrics) -}}
+true
+{{- end -}}
+{{- end -}}
+
+{{/*
+Returns legacy ksm values
+*/}}
+{{- define "newrelic.compatibility.ksm.legacyData" -}}
+enabled: true
+{{- if .Values.kubeStateMetricsScheme }}
+scheme: {{ .Values.kubeStateMetricsScheme }}
+{{- end -}}
+{{- if .Values.kubeStateMetricsPort }}
+port: {{ .Values.kubeStateMetricsPort }}
+{{- end -}}
+{{- if .Values.kubeStateMetricsUrl }}
+staticURL: {{ .Values.kubeStateMetricsUrl }}
+{{- end -}}
+{{- if .Values.kubeStateMetricsPodLabel }}
+selector: {{ printf "%s=kube-state-metrics" .Values.kubeStateMetricsPodLabel }}
+{{- end -}}
+{{- if  .Values.kubeStateMetricsNamespace }}
+namespace: {{ .Values.kubeStateMetricsNamespace}}
+{{- end -}}
+{{- end -}}
+
+{{/*
+Returns the new value if available, falling back on the legacy one
+*/}}
+{{- define "newrelic.compatibility.valueWithFallback" -}}
+{{- if .supported }}
+{{- toYaml .supported}}
+{{- else if .legacy -}}
+{{- toYaml .legacy}}
+{{- end }}
+{{- end -}}
+
+{{/*
+Returns securityContext merged with old runAsUser config
+*/}}
+{{- define "newrelic.compatibility.securityContext" -}}
+{{- if  .Values.runAsUser -}}
+{{- mustMergeOverwrite .Values.securityContext (dict "runAsUser" .Values.runAsUser ) | toYaml }}
+{{- else -}}
+{{- .Values.securityContext | toYaml }}
+{{- end -}}
+{{- end -}}
+
+{{/*
+Returns agent configmap merged with old eventQueueDepth config
+*/}}
+{{- define "newrelic.compatibility.agentConfig" -}}
+{{ $config:= (include "newrelic.compatibility.valueWithFallback" (dict "legacy" .Values.config "supported" .Values.common.agentConfig ) | fromYaml )}}
+{{- if .Values.eventQueueDepth -}}
+{{- mustMergeOverwrite $config (dict "event_queue_depth" .Values.eventQueueDepth ) | toYaml }}
+{{- else -}}
+{{- $config | toYaml}}
+{{- end -}}
+{{- end -}}
```

Comment:
This seems a bit complex to me.
Why not just returning the shimmed yaml?
```suggestion
{{- define "newrelic.compatibility.agentConfig" -}}
{{- with .Values.eventQueueDepth -}}
event_queue_depth: {{ . | quote }}
{{- end -}}
{{- end -}}
```

Reply:
we are merging three things here:
 - .Values.config
 - .Values.common.agentConfig
 - eventQueueDepth

In your example we are just returning if available "event_queue_depth"

Reply:
Moreover we need a reliable function returning the whole agent config since we are checking that in 9-10 places in code:
 - when deciding to create volumes, 
 - when adding the annotation to trigger pod restart

Reply:
I'd say we can still achieve this by merging:
```
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: {{ .Release.Namespace }}
  labels:
    {{- include "newrelic.labels" . | nindent 4 }}
  name: {{ template "newrelic.fullname" . }}-agent
data:
  newrelic-infra.yml: |
    {{- mustMerge .Values.common.agentConfig .Values.config (include "newrelic.compatibility.agentConfig" . | fromYaml) | toYaml | nindent 2 }}
```

The if would still be large, but I still think we can just get rid of it entirely.